### PR TITLE
Added GnuPG and required dependencies

### PR DESCRIPTION
**Pull request by mariusvniekerk**
This is a batch of dependencies that are all needed for GnuPG

### CODE REVIEW

**Comment by mariusvniekerk** [comment]
@jakirkham This is failing with

```
sysutils.o `test -f 'sysutils.c' || echo './'`sysutils.c
sysutils.c: In function ‘gnupg_inotify_watch_socket’:
sysutils.c:1209:52: error: ‘IN_EXCL_UNLINK’ undeclared (first use in this function)
                          (IN_DELETE|IN_DELETE_SELF|IN_EXCL_UNLINK)) == -1)
                                                    ^
sysutils.c:1209:52: note: each undeclared identifier is reported only once for each
```

This seems to be caused by libc being too old?  any suggestions?

**Comment by mariusvniekerk** [comment]
ping @CJ-Wright

**Comment by mariusvniekerk** [comment]
ping @conda-forge/staged-recipes need a bit of help here on libc issues

**Comment by isuruf** [comment]
Maybe add `export CPPFLAGS="-DIN_EXCL_UNLINK=0x04000000"`?

File: recipes/gnupg/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    - ./configure --prefix=$PREFIX
+    - make
+    - make install
+  skip: true  # [win]
+
+requirements:
+  build:
+    - toolchain
+    - gcc
+    - libcxx
```
Comment (isuruf) [inline recipes/gnupg/meta.yaml]: This is unnecessary

**Comment by mariusvniekerk** [comment]
@isuruf that gets a bit further, failing on missing librt now?  Do we package that

**Comment by isuruf** [comment]
Can you post the error message?

**Comment by mariusvniekerk** [comment]
@isuruf fixed it by actively adding `-lrt` to LDFLAGS

File: recipes/gnupg/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    - bzip2
+    - sqlite
+    - libgcrypt
+    - libgpg-error
+    - libksba
+    - libassuan
+    - ntbtls
+    - npth
+  run:
+    - libcxx
```
Comment (isuruf) [inline recipes/gnupg/meta.yaml]: Can you remove `libcxx` from all the recipes?

Comment (mariusvniekerk) [reply recipes/gnupg/meta.yaml]: sure, guess its not needed?

File: recipes/gnupg/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    - libgcc
```
Comment (isuruf) [inline recipes/gnupg/meta.yaml]: If `gfortran` is not used in the package, can you add this only for linux?

File: recipes/gnupg/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    - export LDFLAGS="-lrt $LDFLAGS"                           # [linux]
+    - ./configure --prefix=$PREFIX
+    - make
+    - make install
+  skip: true  # [win]
+
+requirements:
+  build:
+    - toolchain
+    - gcc
```
Comment (isuruf) [inline recipes/gnupg/meta.yaml]: If you don't need `gfortran` add this only for linux

File: recipes/gnupg/meta.yaml (hunk 2 above)
Comment (isuruf) [reply recipes/gnupg/meta.yaml]: If you are not using clang from the clangdev conda package, it's not needed.

Comment (mariusvniekerk) [reply recipes/gnupg/meta.yaml]: basically added this to have the system bit of `check linkages`

Comment (isuruf) [reply recipes/gnupg/meta.yaml]: libc++ being available is a guarantee on OSX >=10.9 and conda-forge packages only support OSX >=10.9

**Comment by mariusvniekerk** [comment]
ping @isuruf

**Comment by mariusvniekerk** [comment]
Doesn't the linter complain about that? Alternatively I can hold off until cb3

File: recipes/gnupg/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    - ./configure --prefix=$PREFIX
+    - make
+    - make install
+  skip: true  # [win]
+
+requirements:
+  build:
+    - toolchain
+    - gcc  # [linux]
+    - bzip2
```
Comment (jakirkham) [inline recipes/gnupg/meta.yaml]: Should be `bzip2 1.0.*`.

File: recipes/gnupg/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    - sqlite
```
Comment (jakirkham) [inline recipes/gnupg/meta.yaml]: Should be `sqlite 3.20.*`.

File: recipes/gnupg/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    - libgcrypt
+    - libgpg-error
+    - libksba
+    - libassuan
+    - ntbtls
+    - npth
+  run:
+    - libgcc  # [linux]
+    - bzip2
+    - sqlite
```
Comment (jakirkham) [inline recipes/gnupg/meta.yaml]: Also need to update pinnings like shown above.

**Comment by jakirkham** [comment]
Are there `make check`s for the different dependencies? Also would quote `$PREFIX`.

**Comment by sodre** [comment]
@mariusvniekerk are you will still working on this? I can help keep it up-to-date as I see a good use for it together with the [git-secret PR](#5467).

**Comment by mariusvniekerk** [comment]
A little,  does the move to cb3 remove the need for explicit pinning, @jakirkham

**Comment by jakirkham** [comment]
At the recipe level, yes. There will be a global pinning file. Re-rendering will use that file to add pinnings and any build matrix. Though some recipes will need to specify how their package should be pinned.

**Comment by mariusvniekerk** [comment]
@jakirkham , given that we have now switched to cb3, do we still need to pin explicitly?

**Comment by jakirkham** [comment]
AFAIK we have to rebuild a lot of things with `run_exports` before pinnings can be dropped. ( https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/27 )

**Comment by mariusvniekerk** [comment]
cc @isuruf restructured this to use the new host build stuff (i think).  Do i need to add run_exports as well?

File: recipes/gnupg/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    - {{ compiler('cxx') }}
+  host:
+    - bzip2
+    - sqlite
+    - libgcrypt
+    - libgpg-error
+    - libksba
+    - libassuan
+    - ntbtls
+    - npth
```
Comment (isuruf) [inline recipes/gnupg/meta.yaml]: Are these all build time dependencies only?

Comment (isuruf) [reply recipes/gnupg/meta.yaml]: I mean aren't they needed at runtime?

Comment (mariusvniekerk) [reply recipes/gnupg/meta.yaml]: adding in some more pins

File: recipes/gnupg/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+  number: 0
+  script:
+    - export CPPFLAGS="-DIN_EXCL_UNLINK=0x04000000 $CPPFLAGS"  # [linux]
+    - export LDFLAGS="-lrt $LDFLAGS"                           # [linux]
+    - ./configure --prefix=$PREFIX
+    - make
+    - make install
+  skip: true  # [win]
+  run_exports:
+    - gnupg
```
Comment (isuruf) [inline recipes/gnupg/meta.yaml]: This should be `- {{ pin_subpackage('zlib', max_pin='x.x') }}` if you assume that there's no ABI changes in patch version

Comment (mariusvniekerk) [reply recipes/gnupg/meta.yaml]: the default is 'x' right? so major

Comment (isuruf) [reply recipes/gnupg/meta.yaml]: `- {{ pin_subpackage('zlib') }}` pins to major. `- gnupg` pins downstream to the package with no version.

Comment (isuruf) [reply recipes/gnupg/meta.yaml]: Edited.

File: recipes/libassuan/meta.yaml [hunk 10]
```diff
@@ -19,7 +19,8 @@ build:
     - make install
   skip: true  # [win]
   run_exports:
-    - libassuan
+    - {{ pin_subpackage('libassuan', max_pin='x.x') }}
```
Comment (dougalsutherland) [inline recipes/libassuan/meta.yaml]: [Looks like](https://abi-laboratory.pro/tracker/timeline/libassuan/) just major should probably be enough.

**Comment by dougalsutherland** [comment]
Shouldn't these be `pin_compatible` and not `pin_subpackage`?

**Comment by isuruf** [comment]
`pin_subpackage` should be used when the package is built in the recipe itself. `pin_compatible` is when you are using a package not in the recipe

**Comment by mariusvniekerk** [comment]
@isuruf this is ready for review now and working.   i added some run_exports bits to libgpg-error and libgcrypt that were already existing.

File: recipes/gnupg/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+  run_exports:
+    - {{ pin_subpackage('gnupg', max_pin='x.x') }}
+
+requirements:
+  build:
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+  host:
+    - bzip2
+    - sqlite
```
Comment (isuruf) [inline recipes/gnupg/meta.yaml]: Is this unneeded in `run`?

File: recipes/gnupg/meta.yaml [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    - libgcrypt
+    - libgpg-error
```
Comment (isuruf) [inline recipes/gnupg/meta.yaml]: Are these two unneeded in run?

Comment (mariusvniekerk) [reply recipes/gnupg/meta.yaml]: both of those have been modified to run_exports

File: recipes/gnupg/meta.yaml (hunk 11 above)
Comment (mariusvniekerk) [reply recipes/gnupg/meta.yaml]: Added

**Comment by mariusvniekerk** [comment]
ping @isuruf @jakirkham

**Comment by jakirkham** [comment]
Sorry @mariusvniekerk. Think @isuruf is swamped with finals. Personally am swamped with conference and other internal work. Let's see if some others can weigh in.

cc @conda-forge/staged-recipes

**Comment by scopatz** [comment]
LGTM, thanks for all of the hard work @mariusvniekerk!

**Pull request merged**